File: packages/app/src/services/software/definitions/python312.ts
import { ok, err } from 'neverthrow';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'python-312';
	public name = 'Python 3.12.x';
	public category = [SoftwareCategory.Development];
	public downloadName = 'python3.12.x-amd64.exe';
	public shouldCacheUrl = true;
	public icon = 'python.png';
	public homepage = 'https://python.org';

	public async resolveDownloadUrl() {
		const res = await fetch('https://python.org/api/v2/downloads/release');
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const releases   = await res.json() as Array<{ name: string; version: number; is_latest: boolean; }>;
		const candidates = releases.filter(r => r.is_latest === false && r.version === 3 && r.name.includes('3.12.'));
		if (candidates.length === 0) {
			return err(DownloadUrlResolveError.Generic);
		}

		const sortedCandidates = candidates.reverse();
		const version          = sortedCandidates[0].name.split(' ')[1];

		return ok(`https://www.python.org/ftp/python/${version}/python-${version}-amd64.exe`);
	}
}

File: src/lib/software/variants/python_312.py
from src.lib.software import BaseSoftware

class Python312(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'python-312'
        self.name = 'Python 3.12.x'
        self.download_name = 'python3.12.x-amd64.exe'
        self.should_cache_url = True
        self.icon = 'python.png'
        self.homepage = 'https://python.org'

        # Releases after 3.12.10 are source code only
        self._version = '3.12.10'

    def resolve_download_url(self):
        self.url_resolved.emit(f'https://www.python.org/ftp/python/{self._version}/python-{self._version}-amd64.exe')
